tests: Bluetooth: bsim: Basic and Mesh test with BT_CTLR_LOW_LAT

cvinayak opened the pull request:
Add tests to cover BT_CTLR_LOW_LAT options which is used by
nRF51 Series.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: basic test with BT_CTLR_LOW_LAT

Add tests to cover BT_CTLR_LOW_LAT options which is used by
nRF51 Series.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Mesh test with BT_CTLR_LOW_LAT

Add tests to cover BT_CTLR_LOW_LAT options which is used by
nRF51 Series.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

trond-snekvik commented on tests/bluetooth/bsim_bt/bsim_test_mesh/_mesh_test.sh:
```diff
@@ -37,7 +37,39 @@ function RunTest(){
 
     echo "Starting $testid as device #$idx"
-    Execute ./bs_${BOARD}_tests_bluetooth_bsim_bt_bsim_test_mesh_prj_conf \
+    Execute \
+      ./bs_${BOARD}_tests_bluetooth_bsim_bt_bsim_test_mesh_prj_conf \
+      -v=${verbosity_level} -s=$1 -d=$idx -RealEncryption=1 \
+      -testid=$testid
+    let idx=idx+1
+  done
+
+  count=$(expr $# - 1 + $extra_devs)
+
+  echo "Starting phy with $count devices"
+
+  Execute ./bs_2G4_phy_v1 -v=${verbosity_level} -s=$1 -D=$count
+
+  for process_id in $process_ids; do
+    wait $process_id || let "exit_code=$?"
+  done
+
+  if [ $exit_code != 0 ]; then
+    exit $exit_code #the last exit code != 0
+  fi
+
+  echo "#### Repeat test for Controller LOW_LAT variant ####"
+  idx=0
+  for testid in ${@:2} ; do
+    if Skip $testid; then
+      echo "Skipping $testid (device #$idx)"
+      let idx=idx+1
+      continue
+    fi
+
+    echo "Starting $testid as device #$idx"
+    Execute \
+      ./bs_${BOARD}_tests_bluetooth_bsim_bt_bsim_test_mesh_prj_low_lat_conf \
```
Instead of this, it'd be better to just add
```sh
conf=${conf:-prj_conf}
Execute ./bs_${BOARD}_tests_bluetooth_bsim_bt_bsim_test_mesh_${conf} \
      -v=${verbosity_level} -s=$1 -d=$idx -RealEncryption=1 \
      -testid=$testid
```
to RunTest(), so we can call
```sh
RunTest mesh_transport_group transport_tx_group transport_rx_group
conf=prj_low_lat_conf
RunTest mesh_transport_group transport_tx_group transport_rx_group
```
in the test scripts, IMO.

The mesh tests generally shouldn't be adding test coverage for the controller, and this makes it impossible to avoid running on two controller variations, which would make debugging of mesh tests a bit of a pain. 

It's just these tests that would benefit from this dual mode testing, IMO:
- transport/unicast.sh
- transport/loopback_group.sh
- friendship/msg_mesh.sh

cvinayak replied:
Will make the suggested change to the RunTest.

> The mesh tests generally shouldn't be adding test coverage for the controller, and this makes it impossible to avoid running on two controller variations, which would make debugging of mesh tests a bit of a pain.

This is not coverage for the controller, it is two controller variants.

aescolar commented on tests/bluetooth/bsim_bt/bsim_test_mesh/_mesh_test.sh:
```diff
@@ -52,7 +54,6 @@ function RunTest(){
   for process_id in $process_ids; do
     wait $process_id || let "exit_code=$?"
   done
-  exit $exit_code #the last exit code != 0
```
The shell script must fail when any device return code is !=0. (otherwise failures won't be reported in CI)
Instead of trying to chain several tests in shell script (like with `tests/bluetooth/bsim_bt/bsim_test_mesh/tests_scripts/transport/unicast.sh`)
maybe add a `unicast_lowlat.sh` and a `msg_mesh_low_lat.sh`

(Apart from the "the test must fail" issue, note that `process_ids` and `exit_code`, (and the possible `SKIP` and `EXTRA_DEVS`) are global for the calling script, so calling Runtest several times is a bad idea)

cvinayak committed:
Bluetooth: Mesh: Update advertising duration calculation

Zephyr Bluetooth Low Energy Controller for mesh stack
uses pre-emptible continuous scanning, allowing advertising
events to be transmitted without delay when advertising is
enabled. No need to compensate with scan window duration.

Zephyr Bluetooth Low Energy Controller built for nRF51x
SoCs use CONFIG_BT_CTLR_LOW_LAT=y, and continuous scanning
cannot be pre-empted, hence, scanning will block advertising
events from being transmitted. Increase the advertising
duration by the amount of scan window duration to compensate
for the blocked advertising events.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Disable scan unreserved for LOW_LAT option

Disable use of BT_CTLR_SCAN_UNRESERVED with BT_CTLR_LOW_LAT.
nRF51x SoC needs to block any CPU use inside radio events,
hence use of radio without time reservation is required for
proper functioning of Controller in nRF51x.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@trond-snekvik  and @jhedberg Need you thoughts on this commit: https://github.com/zephyrproject-rtos/zephyr/pull/35025/commits/f0ad7836972bd4d6680da8641e4e27772152fe8b

Things have changed in the Controller design (nRF51 vs nRF52) in the split architecture, we need to revisit the duration calculation.

trond-snekvik commented on subsys/bluetooth/mesh/adv_legacy.c:
```diff
@@ -53,7 +53,14 @@ static inline void adv_send(struct net_buf *buf)
+	 * This is a calculation specific to Zephyr Bluetooth Low Energy
+	 * Controller when CONFIG_BT_CTLR_LOW_LAT=y used in nRF51x SoCs.
+	 */
+	duration = (BT_MESH_SCAN_WINDOW_MS + adv_int +
```
What's causing this addition of the advertising interval? Does the controller now ensure that stopping and starting the advertiser does not produce any advertising intervals under the specified duration?

cvinayak replied:
In the Controller implementation for nRF51, each radio event prevents any scheduling related CPU use, i.e. inside a scan window event if advertising is enabled then advertising event is scheduled in intervals equivalent to the advertising interval until the event is outside the scan window duration. 

Hence, the first advertising radio event scheduled on air could be delayed upto the duration of the scan window plus one advertising interval.

(please request changes on this PR, to prevent it from being merged until we conclude on the current discussion).

trond-snekvik replied:
I see. Can this be wrapped inside relevant `IS_ENABLED()` checks, so we don't end up sending too many advertisements for the other modes? We're already sending too many for the normal case, to avoid sending none for some corner cases.

trond-snekvik reviewed (approved):
LGTM!
